magma/magma
# Poll Stats GRPC Endpoint Pipelined

Signed-off-by: Vesh Kemburu <veshkemburu@fb.com>



## Summary

Creating a functional pipelined grpc endpoint that replies with stats for a certain subset of rules(matched by cookie mask).
Still in progress.

## Test Plan

1. Add GRPC Endpoint to pipelined.proto, which takes in a GetStatsRequest and returns a RuleRecordTable
2. Define GetStatsRequest message in pipelined.proto to have two attributes: cookie and cookie mask
3. Add argument fields to enforcement_parser in pipelined_cli.py, including a new pull stats command, and two arguments cookie and cookie mask, linking to an rpc wrapper function
4. Define the grpc wrapper function that takes in a client and a set of arguments and returns a response. In this function define a GetStatsRequest using the args, and call GetStats on the client to retrieve a response(then return this response) 
5. In rpc_servicer.py, define GetStats grpc handler. Have this handler invoke the api in flows.py by calling an intermediate function we define in enforcement_stats.py. The intermediate function will return a rule record table which is returned from this function
6. Have the handler in enforcement_stats.py invoke the API? I’m guessing no futures are needed(everything is built into the API), and then use _get_usage_from_flow_stat to convert the result into a RuleRecordTable.
7. From sessiond’s perspective, we need to update PipelinedClient.h to include the new endpoint
8. Add unit test to verify function 

## Additional Information
